File: greedy_algorithm/714_Best Time to Buy and Sell Stock with Transaction Fee.js
/**
 * @param {number[]} prices
 * @param {number} fee
 * @return {number}
 */
var maxProfit = function (prices, fee) {
    const dp = new Array(prices.length).fill([0, 0]);
    dp[0][0] -= prices[0] + fee;//The initial trade comes with a price and a fee

    for (let i = 1; i < prices.length; i++) {
        dp[i][0] = Math.max(dp[i - 1][0], dp[i - 1][1] - prices[i] - fee);//Make sure everytime you buy a stock, the fee comes with it
        dp[i][1] = Math.max(dp[i - 1][1], dp[i - 1][0] + prices[i]);
    }

    return dp[prices.length - 1][1];
};

File: greedy_algorithm/714_Best Time to Buy and Sell Stock with Transaction Fee.cpp
class Solution
{
public:
    int maxProfit(vector<int> &prices, int fee)
    {
        int maxp = 0;
        int min_price = prices[0]; // We buy the first day of stock by default
        for (int i = 1; i < prices.size(); i++)
        {
            // We buy a stock with a lower price, considering the previous stock was sold
            if (prices[i] < min_price)
                min_price = prices[i];

            // We do nothing if the price isn't lower and the profit is negative
            if (prices[i] <= min_price + fee && prices[i] >= min_price)
                continue;

            // We collect profit when it's positive, and update the min_price with the fee because with the fee subtracted, the next real transaction won't be charge for extra fee
            if (prices[i] > min_price + fee)
            {
                maxp += prices[i] - min_price - fee;
                min_price = prices[i] - fee;
            }
        }
        return maxp;
    }
};